security/pkg/pki/ca/ca.go

```diff
@@ -317,9 +329,244 @@ func (ca *IstioCA) SignWithCertChain(csrPEM []byte, subjectIDs []string, ttl tim
+			caSecret.Data[caPrivateKeyID], nil, rootCerts)
+		if err != nil {
+			log.Errorf("failed to create CA KeyCertBundle (%v)", err)
+			return UpgradeFailure
+		}
+
+		if err = updateCACertInConfigmapWithRetry(config, keyCertBundle.GetRootCertPem()); err != nil {
+			log.Errorf("Failed to write self-signed Citadel's root cert "+
+				"to configmap (%s). Node agents will not be able to connect.",
+				err.Error())
+		}
+		ca.setCAKeyCertBundle(keyCertBundle)
+		log.Infof("Updated CA KeyCertBundle using existing public key: %v", string(rootCerts))
+		return UpgradeSuccess
+	}
+	return UpgradeSkip
+}
+
+// shouldRefreshCACert checks CA cert for Self-signed Citadel and informs
+// Citadel to refresh CA cert.
+func (ca *IstioCA) shouldRefreshCACert(config *SelfSignedCARootCertRotationConfig,
+	caSecret *v1.Secret) bool {
+	if caSecret == nil {
+		return false
+	}
+
+	rootCertBytes := caSecret.Data[caCertID]
+	rootCert, err := util.ParsePemEncodedCertificate(rootCertBytes)
+	if err == nil {
+		certLifeTimeLeft := time.Until(rootCert.NotAfter)
+		certLifeTime := rootCert.NotAfter.Sub(rootCert.NotBefore)
+		gracePeriod := time.Duration(config.GracePeriodRatio*1000) * certLifeTime / 1000
+		if certLifeTimeLeft <= gracePeriod {
+			return true
+		}
+	} else {
+		log.Warnf("Failed to parse root cert in secret %s:%s: %s.",
+			caSecret.Namespace, caSecret.Name, err.Error())
+	}
+	return false
+}
+
+func (ca *IstioCA) checkAndRotateRootCertForSigningCertCitadel(
+	config *SelfSignedCARootCertRotationConfig, caSecret *v1.Secret, scrtErr error) RootUpgradeStatus {
+	if scrtErr != nil {
+		log.Errorf("Citadel runs in self-signed mode but"+
+			" fail to load secret %s:%s (error: %s), skip cert rotation job",
+			config.CaStorageNamespace, CASecret, scrtErr.Error())
+		return UpgradeSkip
+	}
+	log.Infof("Self-signed Citadel successfully loaded the secret.")
+	// Check root certificate expiration time in CA secret
+	if ca.shouldRefreshCACert(config, caSecret) {
+		log.Info("Refresh root certificate")
+
+		options := util.CertOptions{
+			TTL:           config.CaCertTTL,
+			SignerPrivPem: caSecret.Data[caPrivateKeyID],
+			Org:           config.Org,
+			IsCA:          true,
+			IsSelfSigned:  true,
+			RSAKeySize:    caKeySize,
+			IsDualUse:     config.DualUse,
+		}
+		pemCert, pemKey, ckErr := util.GenCACertFromExistingKey(options)
+		if ckErr != nil {
+			log.Errorf("unable to generate CA cert and key for self-signed CA: %s", ckErr.Error())
+			return UpgradeFailure
+		}
+
+		rootCerts, err := appendRootCerts(pemCert, config.RootCertFile)
+		if err != nil {
+			log.Errorf("failed to append root certificates: %s", err.Error())
+			return UpgradeFailure
+		}
+
+		keyCertBundle, err := util.NewVerifiedKeyCertBundleFromPem(pemCert, pemKey, nil, rootCerts)
+		if err != nil {
+			log.Errorf("failed to create CA KeyCertBundle (%v)", err)
+			return UpgradeFailure
+		}
+
+		caSecret.Data[caCertID] = pemCert
```

Comment:
Please consider putting the following in a util function in security/pkg/k8s.

Reply:
This part is moved into configmap.go now.